src/ray/common/asio/io_context.cc

```diff
@@ -0,0 +1,257 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/asio/io_context.h"
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <utility>
+
+/// C++11 helper for generalized lambda capture.
+template <typename T, typename F>
+class capture_impl {
+  T x;
+  F f;
+
+ public:
+  capture_impl(T &&x, F &&f) : x{std::forward<T>(x)}, f{std::forward<F>(f)} {}
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) const -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+};
+
+template <typename T, typename F>
+capture_impl<T, F> capture(T &&x, F &&f) {
+  return capture_impl<T, F>(std::forward<T>(x), std::forward<F>(f));
+}
+
+/// A helper for creating a snapshot view of the stats for a handler.
+/// This acquires a reader lock on the provided guarded handler stats, and creates a
+/// lockless copy of the stats.
+HandlerStats to_stats_view(const GuardedHandlerStats &stats) {
+  absl::ReaderMutexLock lock(&(stats.mutex));
```

Comment:
Is Reader/Writer lock meaningful here? I can imagine most of the time we write stats and read-only happens on the main thread once in a while, so it is not necessary? But I might be missing something.

Reply:
My thinking was that there could be readers from multiple threads in the future (inspecting subsets of the stats concurrently), but that was probably premature. I'll change this to a pain lock for now.